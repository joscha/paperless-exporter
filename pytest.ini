[pytest]
python_files = test_*.py
python_functions = test_*
addopts = --show-capture=stdout -vv -s
