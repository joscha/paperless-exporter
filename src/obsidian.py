import hashlib
from pathlib import Path
from shutil import copy
from typing import Dict, Generator

from pathvalidate import sanitize_filename
from .model import (
    DataType,
    ReceiptCollection,
    ReceiptTag,
    Zcategory,
    Zcollection,
    Zpaymentmethod,
    Zreceipt,
    Zsubcategory,
    database,
)
from frontmatter import Post, dump


def get_collection_paths(collection: Zcollection) -> str:
    paths = []
    while collection:
        paths.insert(0, collection.zname)
        collection = collection.parent
    return " / ".join(paths)


def create_out_dir(dir_path: str | Path):
    p = Path(dir_path)
    p.mkdir(parents=True, exist_ok=True)
    return p


def get_document_path(path_to_paperless_db: Path, receipt: Zreceipt):
    return path_to_paperless_db / Path(receipt.zpath)


def get_document_title(receipt: Zreceipt):
    title = receipt.zmerchant
    if not title:
        try:
            title = receipt.zcategory.zname
        except Zcategory.DoesNotExist:
            pass
    if not title:
        title = receipt.zoriginalfilename
    if not title:
        title = f"Paperless document {receipt.z_pk}"
    return title


def get_receipts(path_to_paperless_db: Path) -> Generator[Zreceipt, None, None]:
    database.init(path_to_paperless_db / "DocumentWallet.documentwalletsql")
    try:
        yield from Zreceipt.select()
    finally:
        database.close()


async def export(path_to_paperless_db: Path, out_dir: Path):
    out_dir_path = create_out_dir(out_dir)
    attachments_dir_path = create_out_dir(out_dir_path / "_attachments")

    for receipt in get_receipts(path_to_paperless_db):
        obsidian_item = ObsidianItem(receipt, path_to_paperless_db)
        yield obsidian_item
        obsidian_item.save(out_dir_path, attachments_dir_path)


class ObsidianItem:
    receipt: Zreceipt
    markdown: Post
    path_to_paperless_db: Path

    def __init__(self, receipt: Zreceipt, path_to_paperless_db: Path):
        self.receipt = receipt
        self.path_to_paperless_db = path_to_paperless_db

    def save(self, out_dir_path: Path, attachments_dir_path: Path):
        title = self.get_document_title()
        id = self.receipt.z_pk
        out_file_path = out_dir_path / sanitize_filename(f"{title}.md")
        if out_file_path.exists():
            out_file_path = out_dir_path / sanitize_filename(f"{title} ({id}).md")
            if out_file_path.exists():
                raise Exception(f"File {out_file_path} already exists")

        document_path = self.get_document_path()
        original_document_path = self.get_original_document_path()
        prefix = f"{self.receipt.zdate.strftime("%Y-%m-%d")}_{title}_{id}"

        original_files = {
            "document": document_path,
            "original": original_document_path,
        }
        original_files = {
            k: v for k, v in original_files.items() if v is not None and v.exists()
        }
        if len(original_files) == 0:
            thumbnail_path = self.get_thumbnail_path()
            if thumbnail_path and thumbnail_path.exists():
                original_files["thumbnail"] = thumbnail_path

        linked_attachments = {}
        seen_hashes = set()
        for file_name, file_path in original_files.items():
            file_hash = hashlib.md5(file_path.read_bytes()).hexdigest()
            if file_hash in seen_hashes:
                continue
            seen_hashes.add(file_hash)
            file_out_path = attachments_dir_path / sanitize_filename(
                f"{prefix}.{file_name}{file_path.suffix}"
            )
            copy(file_path, file_out_path)
            linked_attachments[file_name] = file_out_path.relative_to(out_dir_path)
        dump(
            self.transform(linked_attachments=linked_attachments),
            out_file_path,
        )

    def get_document_path(self) -> Path:
        return get_document_path(self.path_to_paperless_db, self.receipt)

    def get_original_document_path(self) -> Path | None:
        if self.receipt.zoriginalfilename:
            document_path = get_document_path(self.path_to_paperless_db, self.receipt)
            document_file_name = document_path.name
            document_file_name_without_extension = document_file_name.rsplit(".", 1)[0]
            original_document_path = (
                document_path.parent
                / document_file_name_without_extension
                / self.receipt.zoriginalfilename
            )
            return original_document_path
        return None

    def get_thumbnail_path(self) -> Path | None:
        if self.receipt.zthumbnailpath:
            return self.path_to_paperless_db / Path(self.receipt.zthumbnailpath)
        return None

    def get_document_title(self) -> str:
        return get_document_title(self.receipt)

    def transform(self, linked_attachments: Dict[str, Path] = None) -> Post:
        receipt = self.receipt
        document_path = self.get_document_path()
        content = []
        if receipt.znotes:
            content.append(receipt.znotes.strip())
            content.append("")
            content.append("-----")
        if linked_attachments and len(linked_attachments) > 0:
            for file_name, file_path in linked_attachments.items():
                content.append(f"#### {file_name}")
                content.append(f"![[{file_path}]]")

        markdown = Post(content="\n".join(content))
        if receipt.zoriginalfilename:
            markdown.metadata["Original filename"] = receipt.zoriginalfilename
        markdown.metadata["Date"] = receipt.zdate.strftime("%Y-%m-%d")
        markdown.metadata["Import date"] = (
            receipt.zimportdate.isoformat()
        )  # .astimezone().replace(microsecond=0).isoformat()
        markdown.metadata["Type"] = receipt.zdatatype.zname

        collections = []
        for receipt_collection in receipt.collections:
            assert isinstance(receipt_collection, ReceiptCollection)
            collections.append(get_collection_paths(receipt_collection.collection))

        if collections:
            markdown.metadata["Collections"] = collections

        try:
            markdown.metadata["Category"] = receipt.zcategory.zname
        except Zcategory.DoesNotExist:
            pass

        try:
            markdown.metadata["Subcategory"] = receipt.zsubcategory.zname
        except Zsubcategory.DoesNotExist:
            pass

        try:
            markdown.metadata["Payment method"] = receipt.zpaymentmethod.zname
        except Zpaymentmethod.DoesNotExist:
            pass

        if receipt.zdatatype.z_pk == DataType.RECEIPT.value:
            markdown.metadata["Amount"] = receipt.zamount
            if receipt.ztaxamount is not None:
                markdown.metadata["Tax/VAT"] = receipt.ztaxamount

        if receipt.zocrattemptedvalue == 1 and receipt.zocrresult:
            markdown.metadata["OCR result"] = receipt.zocrresult

        document_exists = document_path.exists()

        tags = ["paperless"]
        if not document_exists:
            tags.append("missing-document")
        for tag in receipt.receipt_tags:
            assert isinstance(tag, ReceiptTag)
            if tag.tag.zname:
                tags.append(tag.tag.zname.strip().lower())
        if tags:
            markdown.metadata["tags"] = tags

        if document_exists:
            markdown.metadata["source"] = document_path.absolute().as_uri()

        if receipt.zinboxvalue == 1:
            markdown.metadata["In inbox"] = True

        if receipt.zintrashvalue == 1:
            markdown.metadata["In trash"] = True

        return markdown
