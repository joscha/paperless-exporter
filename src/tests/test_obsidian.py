from pathlib import Path
import pytest
from syrupy.filters import props

from . import assert_tree_snapshot

from ..obsidian import (
    CollectionItem,
    ObsidianItem,
    export,
    german_to_ascii,
    unidecode_filename,
)


@pytest.mark.asyncio
async def test_export(tmp_path, snapshot):
    async for item in export(Path("fixtures/library.paperless"), tmp_path):
        if isinstance(item, ObsidianItem):
            assert (
                snapshot(name=f"document_{item.receipt.z_pk}", exclude=props("source"))
                == item.transform(
                    linked_attachments={
                        "document": "linked_attachment.document.pdf",
                        "original": "linked_attachment.original.pdf",
                    }
                ).to_dict()
            )
        elif isinstance(item, CollectionItem):
            assert snapshot(name=f"collection_{item.collection.z_pk}") == item.markdown

    assert_tree_snapshot(tmp_path, snapshot(name="tree"))


def test_unidecode():
    assert german_to_ascii("Geburtstagsgrüße") == "Geburtstagsgruesse"
    assert (
        unidecode_filename(
            "sparhandy 2012-05-25  BASE Internet-Flat (500MB) für effektiv 0,21€ pro Monat inkl. UMTS-Stick - myDealZ.de — myDealZ.de - 25.05.12.pdf"
        )
        == "sparhandy 2012-05-25  BASE Internet-Flat (500MB) fuer effektiv 0,21€ pro Monat inkl. UMTS-Stick - myDealZ.de — myDealZ.de - 25.05.12.pdf"
    )
