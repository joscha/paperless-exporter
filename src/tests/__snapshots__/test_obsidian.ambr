# serializer version: 1
# name: test_export.1
  dict({
    'Category': 'Software/Hardware',
    'Collections': list([
      'Library',
      'My collection',
      'Test Folder / Another collection',
    ]),
    'Date': '2025-04-18',
    'Import date': '2025-04-18T18:19:08.763928+00:00',
    'OCR result': '''
      CPB Software (Germany) GmbH - Im Bruch 3 - 63897 Miltenberg/Main
      Musterkunde AG
      Mr. John Doe
      Musterstr. 23
      12345 Musterstadt Name: Stefanie Müller
      Phone: +49 9371 9786-0
      Invoice WMACCESS Internet
      VAT No. DE199378386
      Invoice No
      Customer No
      Invoice Period
      Date
      123100401
      12345
      01.02.2024 - 29.02.2024
      1. März 2024
      Service Description Amount
      -without VAT- quantity
      Total Amount
      Basic Fee wmView
      130,00 € 1
      130,00 €
      Basis fee for additional user accounts 10,00 € 0
      0,00 €
      Basic Fee wmPos
      50,00 € 0
      0,00 €
      Basic Fee wmGuide
      1.000,00 € 0
      0,00 €
      Change of user accounts
      10,00 € 0
      0,00 €
      Transaction Fee T1
      0,58 € 14
      8,12 €
      Transaction Fee T2 0,70 € 0
      0,00 €
      Transaction Fee T3
      1,50 € 162
      243,00 €
      Transaction Fee T4
      0,50 € 0
      0,00 €
      Transaction Fee T5 0,80 € 0
      0,00 €
      Transaction Fee T6 1,80 € 0
      0,00 €
      Transaction Fee G1 0,30 € 0
      0,00 €
      Transaction Fee G2
      0,30 € 0
      0,00 €
      Transaction Fee G3 0,40 € 0
      0,00 €
      Transaction Fee G4
      0,40 € 0
      0,00 €
      Transaction Fee G5 0,30 € 0
      0,00 €
      Transaction Fee G6 0,30 € 0
      0,00 €
      Total 381,12 €
      VAT 19 % 72,41 €
      Gross Amount incl. VAT 453,53 €
      Terms of Payment: Immediate payment without discount. Any bank charges must be paid by the invoice recipient.
      Bank fees at our expense will be charged to the invoice recipient!
      Please credit the amount invoiced to IBAN DE29 1234 5678 9012 3456 78 | BIC GENODE51MIC (SEPA Credit Transfer)
      This invoice is generated automatically and will not be signed
      Invoice Details
      Period: 01.02.2024 to 29.02.2024
      Unit: Musterkunde AG 12345
      Request sections: T1: T2: T3: T4: T5: T6: G1: G2: G3: G4: G5: G6:
      Amount in Euro: 0,58 0,70 1,50 0,50 0,80 1,80 0,30 0,30 0,40 0,40 0,30 0,30
      user-account-1 10 0 99 0 0 0 0 0 0 0 0 0 154,30 €
      user-account-2 4 0 63 0 0 0 0 0 0 0 0 0 96,82 €
      Transaction Fee Seg T1: T2: T3: T4: T5: T6: G1: G2: G3: G4: G5: G6:
      Queries in Total: 14 0 162 0 0 0 0 0 0 0 0 0
      Total in Euro: 8,12 € 0,00 € 243,00 € 0,00 € 0,00 € 0,00 € 0,00 € 0,00 € 0,00 € 0,00 € 0,00 € 0,00 € 251,12 €
      The explanation of the query fee categories (T1 to T6 and G1 to G6) can be found on our website:
      https://www.wmaccess.com/abfragekategorien
      Invoice Details for wmView Query Reference
      Period: 01.02.2024 to 29.02.2024
      Unit: Musterkunde AG 12345
      wmview, wmProfile and User Profiles Query Segments:
      Query Reference:
      T1: T2: T3: T4: T5: T6:
      *Not specified*
      4 0 9 0 0 0
      AZR/31/27439
      0 0 12 0 0 0
      CCL/3715
      0 0 4 0 0 0
      CRS/28432
      5 0 36 0 0 0
      Cs/52113
      0 0 19 0 0 0
      GS 32090 1 0 7 0 0 0
      Kpi/22695 2 0 6 0 0 0
      PG 7772
      0 0 11 0 0 0
      Rjn/11138
      0 0 15 0 0 0
      SF-M 596/99-08 0 0 5 0 0 0
      Ttrb/17885 1 0 23 0 0 0
      WPN:24791 1 0 4 0 0 0
      Wwt/15658
      0 0 11 0 0 0
      Price for each Query in Euro:
      0,58 0,70 1,50 0,50 0,80 1,80
      15,82 €
      18,00 €
      6,00 €
      56,90 €
      28,50 €
      11,08 €
      10,16 €
      16,50 €
      22,50 €
      7,50 €
      35,08 €
      6,58 €
      16,50 €
    ''',
    'Original filename': 'sample-invoice.pdf',
    'Subcategory': 'My subcategory',
    'Type': 'Email',
    'content': '''
      Some notes to the doc here
      
      -----
      #### document
      ![[linked_attachment.document.pdf]]
      #### original
      ![[linked_attachment.original.pdf]]
    ''',
    'tags': list([
      'paperless',
      'another-tag',
      'my_tag',
    ]),
  })
# ---
# name: test_export.2
  dict({
    'Amount': 93.5,
    'Collections': list([
      'Library',
    ]),
    'Date': '2025-04-18',
    'Import date': '2025-04-18T20:22:35.167018+00:00',
    'In inbox': True,
    'OCR result': '''
      Invoice
      Invoice Number INV-3337
      From:
      DEMO - Sliced Invoices
      Suite 5A-1204
      123 Somewhere Street
      Your City AZ 12345
      admin@slicedinvoices.com
      Order Number 12345
      Invoice Date January 25, 2016
      Due Date January 31, 2016
      Total Due $93.50
      To:
      Test Business
      123 Somewhere St
      Melbourne, VIC 3000
      test@test.com
      Hrs/Qty Service 1.00 Web Design
      This is a sample description... ANZ Bank
      ACC # 1234 1234
      BSB # 4321 432
      Paid
      Rate/Price $85.00 Adjust Sub Total
      0.00% $85.00
      Sub Total $85.00
      Tax $8.50
      Total $93.50
      Payment is due within 30 days from date of invoice. Late payment is subject to fees of 5% per month.
      Thanks for choosing DEMO - Sliced Invoices | admin@slicedinvoices.com
      Page 1/1
    ''',
    'Original filename': 'wordpress-pdf-invoice-plugin-sample.pdf',
    'Payment method': 'Discover',
    'Tax/VAT': 8.5,
    'Type': 'Receipt',
    'content': '''
      #### document
      ![[linked_attachment.document.pdf]]
      #### original
      ![[linked_attachment.original.pdf]]
    ''',
    'tags': list([
      'paperless',
    ]),
  })
# ---
# name: test_export.3
  dict({
    'Collections': list([
      'Library',
    ]),
    'Date': '2025-04-24',
    'Import date': '2025-04-23T22:24:59.945056+00:00',
    'In inbox': True,
    'OCR result': '''
      CPB Software (Germany) GmbH - Im Bruch 3 - 63897 Miltenberg/Main
      Musterkunde AG
      Mr. John Doe
      Musterstr. 23
      12345 Musterstadt Name: Stefanie Müller
      Phone: +49 9371 9786-0
      Invoice WMACCESS Internet
      VAT No. DE199378386
      Invoice No
      Customer No
      Invoice Period
      Date
      123100401
      12345
      01.02.2024 - 29.02.2024
      1. März 2024
      Service Description Amount
      -without VAT- quantity
      Total Amount
      Basic Fee wmView
      130,00 € 1
      130,00 €
      Basis fee for additional user accounts 10,00 € 0
      0,00 €
      Basic Fee wmPos
      50,00 € 0
      0,00 €
      Basic Fee wmGuide
      1.000,00 € 0
      0,00 €
      Change of user accounts
      10,00 € 0
      0,00 €
      Transaction Fee T1
      0,58 € 14
      8,12 €
      Transaction Fee T2 0,70 € 0
      0,00 €
      Transaction Fee T3
      1,50 € 162
      243,00 €
      Transaction Fee T4
      0,50 € 0
      0,00 €
      Transaction Fee T5 0,80 € 0
      0,00 €
      Transaction Fee T6 1,80 € 0
      0,00 €
      Transaction Fee G1 0,30 € 0
      0,00 €
      Transaction Fee G2
      0,30 € 0
      0,00 €
      Transaction Fee G3 0,40 € 0
      0,00 €
      Transaction Fee G4
      0,40 € 0
      0,00 €
      Transaction Fee G5 0,30 € 0
      0,00 €
      Transaction Fee G6 0,30 € 0
      0,00 €
      Total 381,12 €
      VAT 19 % 72,41 €
      Gross Amount incl. VAT 453,53 €
      Terms of Payment: Immediate payment without discount. Any bank charges must be paid by the invoice recipient.
      Bank fees at our expense will be charged to the invoice recipient!
      Please credit the amount invoiced to IBAN DE29 1234 5678 9012 3456 78 | BIC GENODE51MIC (SEPA Credit Transfer)
      This invoice is generated automatically and will not be signed
      Invoice Details
      Period: 01.02.2024 to 29.02.2024
      Unit: Musterkunde AG 12345
      Request sections: T1: T2: T3: T4: T5: T6: G1: G2: G3: G4: G5: G6:
      Amount in Euro: 0,58 0,70 1,50 0,50 0,80 1,80 0,30 0,30 0,40 0,40 0,30 0,30
      user-account-1 10 0 99 0 0 0 0 0 0 0 0 0 154,30 €
      user-account-2 4 0 63 0 0 0 0 0 0 0 0 0 96,82 €
      Transaction Fee Seg T1: T2: T3: T4: T5: T6: G1: G2: G3: G4: G5: G6:
      Queries in Total: 14 0 162 0 0 0 0 0 0 0 0 0
      Total in Euro: 8,12 € 0,00 € 243,00 € 0,00 € 0,00 € 0,00 € 0,00 € 0,00 € 0,00 € 0,00 € 0,00 € 0,00 € 251,12 €
      The explanation of the query fee categories (T1 to T6 and G1 to G6) can be found on our website:
      https://www.wmaccess.com/abfragekategorien
      Invoice Details for wmView Query Reference
      Period: 01.02.2024 to 29.02.2024
      Unit: Musterkunde AG 12345
      wmview, wmProfile and User Profiles Query Segments:
      Query Reference:
      T1: T2: T3: T4: T5: T6:
      *Not specified*
      4 0 9 0 0 0
      AZR/31/27439
      0 0 12 0 0 0
      CCL/3715
      0 0 4 0 0 0
      CRS/28432
      5 0 36 0 0 0
      Cs/52113
      0 0 19 0 0 0
      GS 32090 1 0 7 0 0 0
      Kpi/22695 2 0 6 0 0 0
      PG 7772
      0 0 11 0 0 0
      Rjn/11138
      0 0 15 0 0 0
      SF-M 596/99-08 0 0 5 0 0 0
      Ttrb/17885 1 0 23 0 0 0
      WPN:24791 1 0 4 0 0 0
      Wwt/15658
      0 0 11 0 0 0
      Price for each Query in Euro:
      0,58 0,70 1,50 0,50 0,80 1,80
      15,82 €
      18,00 €
      6,00 €
      56,90 €
      28,50 €
      11,08 €
      10,16 €
      16,50 €
      22,50 €
      7,50 €
      35,08 €
      6,58 €
      16,50 €
    ''',
    'Original filename': 'sample-invoice.pdf',
    'Type': 'Document',
    'content': '''
      #### document
      ![[linked_attachment.document.pdf]]
      #### original
      ![[linked_attachment.original.pdf]]
    ''',
    'tags': list([
      'paperless',
    ]),
  })
# ---
# name: test_export.4
  dict({
    'Collections': list([
      'Library',
    ]),
    'Date': '2025-04-25',
    'Import date': '2025-04-24T22:39:03.893648+00:00',
    'In inbox': True,
    'OCR result': 'Dummy PDF file',
    'Original filename': 'dummy (1).pdf',
    'Type': 'Document',
    'content': '''
      #### document
      ![[linked_attachment.document.pdf]]
      #### original
      ![[linked_attachment.original.pdf]]
    ''',
    'tags': list([
      'paperless',
      'missing-document',
    ]),
  })
# ---
# name: test_export[tree]
  '''
  test_export0/
  ├── _attachments/
  │   ├── 2025-04-18_sample-invoice_1.document.pdf
  │   ├── 2025-04-18_wordpress-pdf-invoice-plugin-sample_2.document.pdf
  │   ├── 2025-04-24_sample-invoice_3.document.pdf
  │   ├── 2025-04-24_sample-invoice_3.original.pdf
  │   └── 2025-04-25_missing_4.thumbnail.png
  ├── missing.md
  ├── sample-invoice (3).md
  ├── sample-invoice.md
  └── wordpress-pdf-invoice-plugin-sample.md
  
  '''
# ---
